canonical/grafana-k8s-operator
# Add integration with oauth

## Issue
Grafana-k8s operator is currently not integrated with IAM stack.
One of its components - Hydra, provides an Oauth relation interface that can be used to integrate charmed applications supporting OIDC/Oauth2 with the IAM bundle and allow to log in with an external identity provider, such as Azure AD.
Grafana is one of the applications that supports both protocols and will therefore be integrated with the IAM bundle as an exemplary third party application in order to create e2e tests.

## Solution
This PR adds integration with hydra-operator through the oauth relation.

## Context
Details of this integration was described in a specification.

Note that a full integration with the IAM stack (including an external identity provider) will be covered in e2e tests in iam-bundle. We currently test on ubuntu 22.04, microk8s 1.26-strict/stable and juju 3.1/stable.

## Testing Instructions
To manually test changes from this PR:
1. Deploy grafana-k8s
2. Deploy postgresql (required dependency for hydra)
`juju deploy postgresql-k8s --channel 14/stable --trust`
3. Deploy hydra, integrate with postgres
`juju deploy hydra --channel edge --trust` (once [this](https://github.com/canonical/hydra-operator/pull/55) PR is merged)
`juju integrate hydra postgresql-k8s`
4. Deploy traefik-public and traefik-admin, relate to hydra:
```
juju deploy traefik-k8s traefik-public --channel edge
juju deploy traefik-k8s traefik-admin --channel edge

juju integrate traefik-public hydra:public-ingress
juju integrate traefik-admin hydra:admin-ingress
```
5. Integrate hydra with grafana:
`juju integrate hydra:oauth grafana-k8s:oauth`
6. If grafana has passed its auth config correctly, `juju show-unit grafana-k8s/0` should return a databag similar to:
```
authorization_endpoint: http://10.64.140.43:80/test3-hydra/oauth2/auth
client_id: 17ea1ab2-6677-48b3-9496-1dc1c70df19b
client_secret_id: secret:<juju secret id>
introspection_endpoint: http://10.64.140.44:80/test3-hydra/admin/oauth2/introspect
issuer_url: http://10.64.140.43:80/test3-hydra
jwks_endpoint: http://10.64.140.43:80/test3-hydra/.well-known/jwks.json
scope: openid profile email phone
token_endpoint: http://10.64.140.43:80/test3-hydra/oauth2/token
userinfo_endpoint: http://10.64.140.43:80/test3-hydra/userinfo
```

If you go to grafana dashboard, you should see a "Log in with Canonical" button, although it will work only if the full [iam bundle](https://github.com/canonical/iam-bundle) is deployed. If you log in with the default admin user, you will be able to see the oauth provider config reflected in the settings.

The flow was tested on our end with Azure AD as the identity provider.
To test the e2e flow on your end, you need to have an external identity provider with kratos registered as an app and provide its details to kratos-external-idp operator.
Note that if you are using self-signed certificates, you need to deploy grafana with an additional variable:
`GF_AUTH_GENERIC_OAUTH_TLS_SKIP_VERIFY_INSECURE=True`, otherwise it will not trust hydra's certificate.


## Release Notes
Added integration with oauth
